src/test/java/com/amazon/dlic/auth/http/jwt/keybyoidc/SelfRefreshingKeySetTest.java

```diff
@@ -11,109 +11,104 @@
-
-        provider.waitForCalled();
+    @Test
+    public void getKey_withKidShouldReturnValidKey() throws AuthenticatorUnavailableException, BadCredentialsException {
 
-        Future<JWK> f2 = executorService.submit(() -> selfRefreshingKeySet.getKey("kid/b"));
+        OctetSequenceKey key = (OctetSequenceKey) selfRefreshingKeySet.getKey("kid/a");
+        assertThat(TestJwk.OCT_1_K, is(equalTo(key.getKeyValue().decodeToString())));
+    }
 
-        while (selfRefreshingKeySet.getQueuedGetCount() == 0) {
-            Thread.sleep(10);
-        }
+    @Test
+    public void getKey_withNullKidShouldThrowAuthenticatorUnavailableException() throws AuthenticatorUnavailableException,
+        BadCredentialsException {
 
-        provider.unblock();
+        Assert.assertThrows(AuthenticatorUnavailableException.class, () -> selfRefreshingKeySet.getKey(null));
 
-        Assert.assertEquals(TestJwk.OCT_1_K, ((OctetSequenceKey) f1.get()).getKeyValue().decodeToString());
-        Assert.assertEquals(TestJwk.OCT_2_K, ((OctetSequenceKey) f2.get()).getKeyValue().decodeToString());
+    }
 
-        Assert.assertEquals(1, selfRefreshingKeySet.getRefreshCount());
-        Assert.assertEquals(1, selfRefreshingKeySet.getQueuedGetCount());
+    @Test
+    public void getKey_withInvalidDataShouldReturnBadCredentialException() throws AuthenticatorUnavailableException,
+        BadCredentialsException {
 
+        Assert.assertThrows(BadCredentialsException.class, () -> selfRefreshingKeySet.getKey("kid/X"));
     }
 
-    static class MockKeySetProvider implements KeySetProvider {
-
-        @Override
-        public JWKSet get() throws AuthenticatorUnavailableException {
-            return TestJwk.OCT_1_2_3;
-        }
+    @Test
+    public void getKeyAfterRefresh_withKidShouldReturnKey() throws AuthenticatorUnavailableException, BadCredentialsException {
 
+        OctetSequenceKey key = (OctetSequenceKey) selfRefreshingKeySet.getKeyAfterRefresh("kid/b");
+        assertThat(TestJwk.OCT_2_K, is(equalTo(key.getKeyValue().decodeToString())));
     }
 
-    static class BlockingMockKeySetProvider extends MockKeySetProvider {
-        private boolean blocked = true;
-        private boolean called = false;
+    @Test
+    public void getKeyAfterRefresh_queuedGetCountVariableShouldBeZeroWhenFinishWithAllKeyRefreshes() throws InterruptedException,
+        ExecutionException {
+
+        int numThreads = 10;
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+        Object lock = new Object();
+
+        for (int i = 0; i < numThreads; i++) {
+            executor.submit(() -> {
+                synchronized (lock) {
+                    try {
+                        selfRefreshingKeySet.getKeyAfterRefresh("kid/a");
+                    } catch (AuthenticatorUnavailableException e) {} catch (BadCredentialsException e) {}
+                }
+            });
+        }
```

Comment:
I'm not sure how this works or what it tests, can you provide more details.

Also of note;
- Synchronize block at @78 prevents the threads from executing at the same time
- the method `getKeyAfterRefresh(...)` is synchronized, so it cannot be executed in parallel, even without the sync block

Reply:
@peternied you're correct. The lock object is preventing the threads to execute at the same time. However, my intention was to make sure that getRefreshCount() has the same amount of calls that we have in numThreads and getQueuedGetCount() has zero pending calls.
I created another test without lock and the assertions are quite different since there is not lock object, so the main thread just keep executing. I'll push it and see what you think.

Reply:
@peternied, while working on this code I found out that this test runs as synchronous code, so all of that code doesn't do much. I'll remove it since it doesn't have any insights at all :disappointed:.
Thanks a lot for the feedback, it helped me to understand better how asynchronous calls work, and also to discover that testing race condition is really, really hard.

Reply:
> I'll remove it since it doesn't have any insights at all 😞. what do you think?

I'd rather remove it - that way no one is under the false impression that something is tested better than it really is.

> testing race condition is really, really hard.

🤣 Indeed!